tracing-subscriber/src/filter/directive.rs

```diff
@@ -0,0 +1,364 @@
+                    "invariant violated: Ordering::Equal must imply a.field_names == b.field_names"
+                );
+            }
+        }
+
+        ordering
+    }
+}
+
+impl PartialOrd for StaticDirective {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Match for StaticDirective {
+    fn cares_about(&self, meta: &Metadata<'_>) -> bool {
+        // Does this directive have a target filter, and does it match the
+        // metadata's target?
+        if let Some(ref target) = self.target {
+            if !meta.target().starts_with(&target[..]) {
+                return false;
+            }
+        }
+
+        if meta.is_event() && !self.field_names.is_empty() {
+            let fields = meta.fields();
+            for name in &self.field_names {
+                if fields.field(name).is_none() {
+                    return false;
+                }
+            }
+        }
+
+        true
+    }
+
+    fn level(&self) -> &LevelFilter {
+        &self.level
+    }
+}
+
+impl Default for StaticDirective {
+    fn default() -> Self {
+        StaticDirective {
+            target: None,
+            field_names: FilterVec::new(),
+            level: LevelFilter::ERROR,
+        }
+    }
+}
+
+impl fmt::Display for StaticDirective {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut wrote_any = false;
+        if let Some(ref target) = self.target {
+            fmt::Display::fmt(target, f)?;
+            wrote_any = true;
+        }
+
+        if !self.field_names.is_empty() {
+            f.write_str("[")?;
+
+            let mut fields = self.field_names.iter();
+            if let Some(field) = fields.next() {
+                write!(f, "{{{}", field)?;
+                for field in fields {
+                    write!(f, ",{}", field)?;
+                }
+                f.write_str("}")?;
+            }
+
+            f.write_str("]")?;
+            wrote_any = true;
+        }
+
+        if wrote_any {
+            f.write_str("=")?;
+        }
+
+        fmt::Display::fmt(&self.level, f)
+    }
+}
+
+impl FromStr for StaticDirective {
+    type Err = ParseError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut split = s.split('=');
+        let part0 = split
+            .next()
+            .ok_or_else(|| ParseError::msg("string must not be empty"))?;
+
+        if let Some(part1) = split.next() {
+            let mut split = part0.split("[{");
+            let target = split.next().map(String::from);
+            let mut field_names = FilterVec::new();
+            if let Some(maybe_fields) = split.next() {
+                let fields = maybe_fields
+                    .strip_suffix("}]")
+                    .ok_or_else(|| ParseError::msg("expected fields list to end with '}]'"))?;
+                field_names.extend(fields.split(',').filter_map(|s| {
```

Comment:
Hm, is this `.split(',')` correct? There's a `.split(',')` which the `Target` does: 

https://github.com/matklad/tracing/blob/33f030f3f6a12ff6543e5969eea3ca549ab4dc7c/tracing-subscriber/src/filter/targets.rs#L257-L259

So, the two would conflict I believe? 


With my "stuck writing parsers" hat on, I sort-of feel that neither regexes, nor `.split` are the most appropriate parsing tools in this case. I personally would reach for "let's write recursive descent parser by hand" here, sort-of how sevmer does this: https://github.com/dtolnay/semver/blob/master/src/parse.rs.

Reply:
You're right that as this is _currently_ used, there will never be commas inside the directive, because this `FromStr` impl is only called in `Targets::from_str`, which already splits on commas. However, I wanted to handle directives with field lists in this `FromStr` impl, because I thought it would eventually be nice to allow the user to manually parse individual `StaticDirectives`. Also, if people want to use field *name* filtering with `Targets`, we could eventually rewrite the `Targets` parser to not just split on commas and use recursive descent or something, so that commas can be nested within filter directives.

I wanted to get something simple working now, though, and go back and add support for field name filtering later.